snap/integrity/integrity.go

```
@@ -0,0 +1,284 @@
+ *
+ */
+
+package integrity
+
+import (
+	"bytes"
+	"crypto"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap/integrity/dmverity"
+	"github.com/snapcore/snapd/snap/squashfs"
+)
+
+const (
+	blockSize = 4096
+	// For now that the header only includes a fixed-size string and a fixed-size hash,
+	// the header size is always gonna be less than blockSize and will always get aligned
+	// to blockSize.
+	HeaderSize = 4096
+	// https://github.com/plougher/squashfs-tools/blob/master/squashfs-tools/squashfs_fs.h#L289
+	squashfsSuperblockBytesUsedOffset = 40
+)
+
+var (
+	// magic is the magic prefix of snap extension blocks.
+	magic = []byte{'s', 'n', 'a', 'p', 'e', 'x', 't'}
+)
+
+// align aligns input `size` to closest `blockSize` value
+func align(size uint64) uint64 {
+	return (size + blockSize - 1) / blockSize * blockSize
+}
+
+// An IntegrityData structure represents a snap's integrity data and
+// contains information useful for locating it.
+type IntegrityData struct {
+	Header         *IntegrityDataHeader
+	Offset         uint64
+	SourceFilePath string
+}
+
+// FindIntegrityData returns integrity data information given a snap filename.
+// It currently supports only integrity data attached at the end of snap files.
+func FindIntegrityData(snapName string) (*IntegrityData, error) {
+	integrityData := IntegrityData{}
+
+	snapFile, err := os.OpenFile(snapName, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer snapFile.Close()
+
+	snapFileInfo, err := snapFile.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	if !squashfs.FileHasSquashfsHeader(snapName) {
+		return nil, fmt.Errorf("input file does not contain a SquashFS filesystem")
+	}
+
+	// Seek to bytes_used field of SquashFS superblock
+	_, err = snapFile.Seek(squashfsSuperblockBytesUsedOffset, io.SeekStart)
```

this is ignoring error

---

Fixed.